# The main project 

__import libraries:__
below are all the libraries used in this project

In [51]:
%matplotlib inline 
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import datetime
import seaborn as sns
#TODO
import sklearn
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score, recall_score, precision_recall_curve,f1_score, fbeta_score, confusion_matrix


__Reading the file and its properties__


In [40]:
df = pd.read_csv('project_dataset.csv') 
print('shape is {}: '.format(df.shape))
print(df.dtypes)
print('\n')
df.sample(10)

shape is (119390, 32): 
hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent     

hotel  is_canceled  lead_time  arrival_date_year  \
110380    City Hotel            0         41               2017   
86551     City Hotel            0          6               2016   
32352   Resort Hotel            0          4               2017   
91916     City Hotel            0        166               2016   
76665     City Hotel            0          6               2015   
91404     City Hotel            0         14               2016   
26660   Resort Hotel            0        114               2016   
82156     City Hotel            1         99               2016   
14652   Resort Hotel            0        266               2015   
114711    City Hotel            0        150               2017   

       arrival_date_month  arrival_date_week_number  \
110380              April                        17   
86551               April                        15   
32352             January                         4   
91916                June                        27   
76665            November                        47   
91404                June                        26   
26660              August                        32   
82156            February                         8   
14652           September                        39   
114711               June                        25   

        arrival_date_day_of_month  stays_in_weekend_nights  \
110380                         26                        0   
86551                           3                        2   
32352                          22                        2   
91916                          26                        1   
76665                          17                        0   
91404                          19                        2   
26660                           6                        0   
82156                          19                        0   
14652                          24                        2   
114711                         19                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
110380                     1       1  ...    No Deposit    NaN   280.0   
86551                      1       1  ...    No Deposit   83.0     NaN   
32352                      0       1  ...    No Deposit    6.0     NaN   
91916                      0       2  ...    No Deposit    9.0     NaN   
76665                      2       1  ...    No Deposit    NaN    40.0   
91404                      0       2  ...    No Deposit   14.0     NaN   
26660                      1       2  ...    No Deposit  240.0     NaN   
82156                      1       2  ...    No Deposit    NaN    67.0   
14652                      4       2  ...    No Deposit  201.0     NaN   
114711                     5       2  ...    No Deposit    7.0     NaN   

       days_in_waiting_list    customer_type     adr  \
110380                    0  Transient-Party  110.00   
86551                     0        Transient   71.20   
32352                     0        Transient   37.00   
91916                     0  Transient-Party   96.30   
76665                     0        Transient   65.00   
91404                     0        Transient  144.20   
26660                     0        Transient  189.00   
82156                     0  Transient-Party   80.00   
14652                     0  Transient-Party   45.45   
114711                    0        Transient   85.59   

        required_car_parking_spaces  total_of_special_requests  \
110380                            0                          0   
86551                             0                          1   
32352                             1                          0   
91916                             0                          1   
76665                             0                          0   
91404                             0                          1   
26660                             0                          1   
82156                       

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [42]:

df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

__there is something here__
there is a lot of thing here must to be considerd and remove such as the company and the agent which are something that we don't want and containg a lot of null values, and also i will compine the date of the first cople of cloumn and maKe it one with datetime data type

# Cleaning Part

In [43]:
df = df[~(df.children.isna())]
# by this filter we will cut 4 rows.


df['assigned_reserved_room'] = np.where(df.assigned_room_type == df.reserved_room_type, 1, 0) # 1 means the assigning room is the same as reserved room.
df.arrival_date_month = df.arrival_date_month.apply(lambda x: datetime.datetime.strptime(x, "%B").month) # convert the month name to its number

df['date'] = df[['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month']].apply(lambda x: '-'.join(x.values.astype(str)), axis="columns")# merge three columns into one
df['date'] = pd.to_datetime(df['date']) #convert it into datetime

df.children = df.children.astype(int)#convert children columns from float to int

df = df.drop(['company', 'agent', 'assigned_room_type', 'reserved_room_type','arrival_date_week_number', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month'], axis=1) # drop not useful columns

Now i will remove the data that i thinK it's will be outliers and too far from the others.

 * first i will checK the adults column becuse there are alot of numbers that does not maKe since such as 55. 50,40,24,26 and 20 all of them were have one record so i thinK it's for testing or for somthing liKe that after that
 * and the same thing with the children one of the record was 10 children and i believe this is not correct
 * also the same with babies twice 9 and 10 babies were recorded

 and we will lose 17 rows in total

In [44]:
firstMask = df.adults < 6 # because there is only 14 adult how has more than 6
secondMask = df.children < 6 # because there is only 1 children how has more than 6
thirdMask = df.babies < 6 # because there is only 2 babies how has more than 6
df = df[firstMask & secondMask & thirdMask]

### now the data locKs better now and i thing we can start chosing the best model.
but before that lets checK if the target data is balanced or not

In [45]:
df.groupby('is_canceled').count()

hotel  lead_time  stays_in_weekend_nights  stays_in_week_nights  \
is_canceled                                                                    
0            75164      75164                    75164                 75164   
1            44205      44205                    44205                 44205   

             adults  children  babies   meal  country  market_segment  ...  \
is_canceled                                                            ...   
0             75164     75164   75164  75164    74743           75164  ...   
1             44205     44205   44205  44205    44138           44205  ...   

             deposit_type  days_in_waiting_list  customer_type    adr  \
is_canceled                                                             
0                   75164                 75164          75164  75164   
1                   44205                 44205          44205  44205   

             required_car_parking_spaces  total_of_special_requests  \
is_canceled                                                           
0                                  75164                      75164   
1                                  44205                      44205   

             reservation_status  reservation_status_date  \
is_canceled                                                
0                         75164                    75164   
1                         44205                    44205   

             assigned_reserved_room   date  
is_canceled                                 
0                             75164  75164  
1                             44205  44205  

[2 rows x 25 columns]

# model problem and need the data to be clean

- the first problem was the spaceing the in the column names.

By this information we can see that the there are 75164 not canceled and 44205 is canceled which lead to the answer of the above quetion 

# Model

Now I will define the target value and the data and since there is problem with the uniqueness of the data, and when I implied the duplicated mtheod i notice that there are arround 33% of the data were duplicated and for that i notice other thing which is there is no id or uniqe vlaue to specifiy each record and i thinK the duplicated happed because there might some people have the same value but that does not mean they are duplacated. and for that i will create another dataset for the duplacted value and i will test it in the final to see the diffrence between them.



In [46]:
# data without duplicated
no_duplicated_data = df.drop_duplicates()
data_nodup = no_duplicated_data.drop(columns=['is_canceled'], axis=1)
targer_nodup = no_duplicated_data['is_canceled']

#splotting the data
data = df.drop(columns = ['is_canceled'], axis=1)
target = df['is_canceled']
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=1) # the traing size will be 80% while the test size is 20%
print(data.shape)
print(target.shape)
print(X_train.shape)
print(X_test.shape)

(119369, 25)
(119369,)
(95495, 25)
(23874, 25)


After splitting the data we will try for every classifcation model and test the scores and see what is the best score for this model.

## Evaluation procedure K-nearest Neighbors (KNN) #1:

In [47]:
data.head(4)

hotel  lead_time  stays_in_weekend_nights  stays_in_week_nights  \
0  Resort Hotel        342                        0                     0   
1  Resort Hotel        737                        0                     0   
2  Resort Hotel          7                        0                     1   
3  Resort Hotel         13                        0                     1   

   adults  children  babies meal country market_segment  ... deposit_type  \
0       2         0       0   BB     PRT         Direct  ...   No Deposit   
1       2         0       0   BB     PRT         Direct  ...   No Deposit   
2       1         0       0   BB     GBR         Direct  ...   No Deposit   
3       1         0       0   BB     GBR      Corporate  ...   No Deposit   

   days_in_waiting_list  customer_type   adr  required_car_parking_spaces  \
0                     0      Transient   0.0                            0   
1                     0      Transient   0.0                            0   
2                     0      Transient  75.0                            0   
3                     0      Transient  75.0                            0   

  total_of_special_requests  reservation_status reservation_status_date  \
0                         0           Check-Out              2015-07-01   
1                         0           Check-Out              2015-07-01   
2                         0           Check-Out              2015-07-02   
3                         0           Check-Out              2015-07-02   

   assigned_reserved_room       date  
0                       1 2015-07-01  
1                       1 2015-07-01  
2                       0 2015-07-01  
3                       1 2015-07-01  

[4 rows x 25 columns]

# Here